File: SparseMatrix.m
% ****************************************
%   Metodi del Calcolo Scientifico 2016
% ****************************************

clear all
clc

% ottengo la lista dei file .mat nella cartella
folderName = input('type folder name: ','s')
matfolder = what(folderName);
addpath(matfolder.path)
matlist = matfolder.mat;

for i = 1:(length(matlist))
    
    disp(['======================================']);

    % carico il file di dati ed ottengo il referirmento alla matrice dei
    % coefficienti A
    mat = load(char(matlist(i))); 
    A = mat.Problem.A; 
    
    disp(['Name matrix: ', mat.Problem.name]);
    disp(['Rows: ', num2str(size(A,2))]);
    disp(['Columns: ', num2str(size(A,1))]);
    disp(['Nonzero: ', num2str(nnz(A))]);
    disp(['-  -  -  -  -  -  -  -  -  -']);

    % abilito il profiler all'analisi della memoria
    profile clear
    profile -memory -history on

    % ottengo il vettore colonna unitario con numero righe di A
    xe = ones(length(A),1);
    % ottengo il vettore dei termini noti b = A * xe
    b = A * xe;
    
    tic
    % risoluzione del sistema lineare (con UMFPACK)
    x = A\b; 
    t = toc;
    disp(['Solving time: ',num2str(t),' s']);

    % calcolo errore relativo
    relErr = norm(x-xe)/norm(xe);
    disp(['Relative error: ',num2str(relErr)]);

    % genero il report dell'occupazione della memoria
    %profile viewer
    p = profile('info');
    mem = (p.FunctionTable(4).TotalMemAllocated)/1000;
    disp(['Allocated memory: ',num2str(mem),' kB']);
    
    % memorizzo i dati ottenuti in una lista
     listMatrix(i).name_matrix = mat.Problem.name;
     listMatrix(i).rows = size(A,2);
     listMatrix(i).solv_time = t;
     listMatrix(i).relative_error = relErr;
     listMatrix(i).alloc_memory = double(mem);
    
end

sortrows(struct2table(listMatrix),2)

File: proj1-Python/SparseMatrix.py

# ****************************************
#   Metodi del Calcolo Scientifico 2016
# ----------------------------------------
#   743464 Banfi Alessandro
#   735722 Curatitoli Mattia
#   742752 Ghelfi Camillo
# ****************************************

# *****************************************************
# Per una corretta esecuzione installare i pacchetti:
# - scikit-umfpack (0.2.1)
# - memory_profiler (0.41)
# - psutil (4.2.0)
# *****************************************************

# package per interfacciarsi col sistema operativo
import os as os
# package scipy con moduli per calcoli matematici
import scipy as scipy 
# modulo input/output files
import scipy.io as scipyio
# modulo matrici Compressed Sparse Column
from scipy.sparse import csc_matrix 
# modulo per algebra lineare
from scipy.sparse import linalg 
# modulo per calcolo della norma in matrici sparse
from scipy.linalg import norm 
# modulo per il timestamp
from datetime import datetime 
# modulo profiler memoria
from memory_profiler import profile
import psutil

# file di log della memoria rilevata dal profiler
# mem_prof = open('memory_profiler.log','w+')
# @profile(stream=mem_prof)
@profile
def xSolver(A,b,bool):
	x = scipy.sparse.linalg.spsolve(A,b,use_umfpack=bool)
	return x


def matrixSolver(folder, eachmtx):
	print '======================================'
	# ottengo il percorso di ogni file .mtx
	pathmtx = folder + '/' + eachmtx
	# leggo le info del file .mtx con scipyio.mminfo()
	mminfo = scipyio.mminfo(pathmtx)
	# scipyio.mminfo() = [rows, cols, entries, format, field, symmetry]
	# -> mminfo[0] = numero righe della matrice
	rows = mminfo[0] 
	# carico file .mtx e converto in matrice csc (Compressed Sparse Column) 
	A = scipyio.mmread(pathmtx).tocsc()
	# ottengo elementi diversi da zero in A
	nonzero = A.nnz

	print 'Name matrix: ' + pathmtx
	print 'Rows: ' + str(rows)
	print 'Columns: ' + str(rows)
	print 'Nonzero: ' + str (nonzero)
	print '-  -  -  -  -  -  -  -  -  -'

	# ottengo il vettore colonna unitario xe con numero righe di A
	xe = scipy.ones(rows) 
	# ottengo il vettore dei termini noti b = A * xe
	b = A * xe 
	# for i in range(0,10):
	# 	print format(b[i], '.30f')
	
	# inizializzo il vettore soluzione con il numero righe di A
	x = scipy.empty(rows) 
	timerOn = datetime.now() 
	# risoluzione del sistema lineare senza UMFPACK
	x = xSolver(A,b,False)
	t = datetime.now() - timerOn
	print 'Solving time without UMFPACK: ' + str(t) + ' s'

	# reinizializzo il vettore soluzione con il numero righe di A
	x = scipy.empty(rows) 
	timerOn = datetime.now() 
	# risoluzione del sistema lineare con UMFPACK
	x = xSolver(A,b,True)
	t = (datetime.now() - timerOn)
	print 'Solving time with UMFPACK: ' + str(t) + ' s'

	# calcolo errore relativo
	relErr = norm(x-xe)/norm(xe)
	print 'Relative error: ' + str(relErr)


def loopSparseMatrix(f):
	folder = f
	if os.path.isdir(folder): 
		path = os.path.abspath(folder)
		# elenco tutti i file .mtx contenuti nella cartella indicata in 'folder'
		mtxlist = [each for each in os.listdir(path) if each.endswith('.mtx')]

		for eachmtx in mtxlist:

			matrixSolver(folder, eachmtx)


if __name__ == '__main__':
  # funzione per la soluzione di tutte le matrici nella cartella indicata
  loopSparseMatrix('FEMLAB')
  
  # funzioni per la soluzione di ciascuna singola matrice
  # matrixSolver('FEMLAB', 'ns3Da.mtx')
  # matrixSolver('FEMLAB', 'poisson2D.mtx')
  # matrixSolver('FEMLAB', 'poisson3Da.mtx')
  # matrixSolver('FEMLAB', 'poisson3Db.mtx')
  # matrixSolver('FEMLAB', 'problem1.mtx')
  # matrixSolver('FEMLAB', 'sme3Da.mtx')
  # matrixSolver('FEMLAB', 'sme3Db.mtx')
  # matrixSolver('FEMLAB', 'sme3Dc.mtx')


